### PR TITLE
Add a function to download PDFs via a browser

### PR DESCRIPTION
**Pull request by adomasven**

### CODE REVIEW

File: chrome/content/zotero/xpcom/attachments.js [hunk 1]
```diff
@@ -1105,9 +1106,91 @@ Zotero.Attachments = new function(){
...
+				enforcingPDF = true;
 				await _enforcePDF(path);
 			}
 		}
+		catch (e) {
+			try {
+				await OS.File.remove(path, { ignoreAbsent: true });
+			}
+			catch (e) {
+				Zotero.debug(e, 1);
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/attachments.js:1118]: → `Zotero.logError(e)`

File: chrome/content/zotero/xpcom/attachments.js [hunk 2]
```diff
@@ -1105,9 +1106,91 @@ Zotero.Attachments = new function(){
...
+			}
+			// Custom handling for PDFs that are bot-guarded
+			// via a JS-redirect
+			if (enforcingPDF && e instanceof this.InvalidPDFException
+					&& url.startsWith('https://www.sciencedirect.com/')) {
+				this.downloadPDFViaBrowser(url, path, options);
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/attachments.js:1124]: `await`

File: test/tests/attachmentsTest.js [hunk 3]
```diff
@@ -279,6 +279,29 @@ describe("Zotero.Attachments", function() {
 		});
 	});
 	
+	
+	describe("#importFromURL()", function () {
+		it("should download a PDF from a JS redirect page", async function () {
+			this.timeout(65e3);
+			
+			let stub = sinon.stub(Zotero.Attachments, 'downloadFile');
+			stub.callsFake(Zotero.Attachments.downloadPDFViaBrowser.bind(Zotero.Attachments));
```
Comment (dstillman) [inline test/tests/attachmentsTest.js:288]: We should just hard-code our test URL above so we don't have to fake this (which probably caused this to miss the missing `await` in `downloadFile()`).

File: test/tests/attachmentsTest.js [hunk 4]
```diff
@@ -279,6 +279,29 @@ describe("Zotero.Attachments", function() {
...
+
+			var item = await Zotero.Attachments.importFromURL({
+				libraryID: Zotero.Libraries.userLibraryID,
+				url: 'https://zotero-static.s3.amazonaws.com/test-pdf-redirect.html',
+				contentType: 'application/pdf'
+			});
+			
+			assert.isTrue(item.isPDFAttachment());
```
Comment (dstillman) [inline test/tests/attachmentsTest.js:296]: We should test that the PDF was actually saved to disk (by stealing a couple lines from `_enforcePDF()`).

File: chrome/content/zotero/xpcom/attachments.js [hunk 5]
```diff
@@ -1105,9 +1106,91 @@ Zotero.Attachments = new function(){
...
+	 */
+	this.downloadPDFViaBrowser = async function (url, path, options = {}) {
+		Zotero.debug(`downloadPDFViaBrowser: Downloading file via browser from ${url}`);
+		const timeout = 60e3;
+		let channelBrowser, hiddenBrowser;
+		let hiddenBrowserPDFFoundDeferred = Zotero.Promise.defer();
+		
+		var pdfMIMETypeHandler = async (data, name, _, channel) => {
+			Zotero.debug(`downloadPDFViaBrowser: Sniffing a PDF loaded at ${name}`);
+			Zotero.debug(data.substr(0, 10));
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/attachments.js:1146]: We can remove this — no need to dump binary data to the terminal.

File: chrome/content/zotero/xpcom/attachments.js [hunk 6]
```diff
@@ -1105,9 +1106,91 @@ Zotero.Attachments = new function(){
...
+			try {
+				channelBrowser = channel.notificationCallbacks.getInterface(Ci.nsIWebNavigation)
+					.QueryInterface(Ci.nsIDocShell).chromeEventHandler;
+			}
+			catch (e) {}
+			if (channelBrowser) {
+				isOurPDF = hiddenBrowser === channelBrowser;
+			}
+			else {
+				// try the document for the load group
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/attachments.js:1159]: Can you clarify when this would be relevant?

Comment (adomasven) [reply chrome/content/zotero/xpcom/attachments.js:1159]: I have no clue. This is taken from, but there must be a reason. Git blame does not reveal it though. https://github.com/zotero/zotero/blob/61e976bd3aec95b73cd993e685adfa897ab3784f/chrome/content/zotero/xpcom/cookieSandbox.js#L331-L360

**Comment by adomasven** [comment]
Also added a bit of code to deproxify URLs.

**Comment by dstillman** [comment]
This saved an invalid PDF for me, containing UTF-8 replacement characters (`EF BF BD`).

https://www.sciencedirect.com/science/article/pii/S0968090X21004307

**Comment by adomasven** [comment]
Updated

**Comment by dstillman** [comment]
Oh, I missed that a test is failing after this: "should download open-access PDF if a translated PDF fails"

**Pull request merged**